### ISSUE TITLE
Create branch policy in the main repo

### ISSUE THREAD

**Issue by kparzysz-quic** [llvm/llvm-project#56643]
Decide whether we allow developers to create branches in the main repository.  If so, establish the policy for maintaining these branches.

**Comment by kparzysz-quic** [comment]
The outcome of this will shape the https://github.com/llvm/llvm-project/issues/56636.

**Comment by rengolin** [comment]
First things first, I'm strongly against development branches in a tree that we only forward rebase, in a project that we all commit to upstream-first incremental development, so I'd be against anything that creates branches for more than just stacked PRs.

For the sole purpose of stacked PRs, I believe GH allows you the option to delete the branch once the PR is merged, so that's self-cleaning in a way.

And if we make that a strong polity (that no branch can exist other than for stacked PRs), we can easily create a script that removes any branch in the main repo that isn't attached to a PR, and run that regularly.

Of course, if GH has an option like that, then we should just enable it.

**Comment by kparzysz-quic** [comment]
I think the policy should address two main topics:
1. Creation/maintenance of topic branches.
2. Merging topic branches into development branches (`main`, `release/xyx`, etc).

**Creation/maintenance of topic branches**

As I see it, there are two main options:
1. Allow developers to create topic branches in llvm-project. The downside of the first option is that this can lead to hundreds of leftover branches, which is messy. This should also include policy on deleting stale (or otherwise unneeded) branches.
2. Do not allow topic branches, and only allow cross-fork pull requests. The downside of the second option is that it would prevent chaining PRs. This may or may not be a big deal, depending on whether we decide to use chained PRs or not.

**Merging topic branches**

First of all, there is still the question of users merging branches into main directly, without GitHub.  The move to PRs may seem to imply that direct pushes are no longer allowed, but I wanted to address that explicitly.

For merging of PRs, GitHub allows three options:
1. Create a merge commit:  this option makes the commit history non-linear.  If we follow the pre-existing policy of keeping the history linear, we should disallow it.
2. Rebase and merge: rebase the topic branch associated with the PR on top of the base branch of the PR. Every commit in the PR branch will be present as a separate commit in the base branch.
3. Squash and merge: squash all commits in the PR branch into one commit and cherry-pick to the base branch.

**Comment by kparzysz-quic** [comment]
> First things first, I'm strongly against development branches in a tree that we only forward rebase, in a project that we all commit to upstream-first incremental development, so I'd be against anything that creates branches for more than just stacked PRs.
> 
> For the sole purpose of stacked PRs, I believe GH allows you the option to delete the branch once the PR is merged, so that's self-cleaning in a way.
> 
> And if we make that a strong policy (that no branch can exist other than for stacked PRs), we can easily create a script that removes any branch in the main repo that isn't attached to a PR, and run that regularly.
> 
> Of course, if GH has an option like that, then we should just enable it.

GitHub can auto-delete merged branches, I think that enabling this option would be a good idea in any case.

The problem with only allowing topic branches when they are used in chained PRs is that once all but 1 PR remains unmerged, the remaining PR is no longer "stacked".  In that sense, a "standalone" PR is a special case of chained PRs.

GitHub will not allow a branch to be deleted if there is an open PR for this branch anyway, so I propose an amendment to your idea, that we only delete (via scripts) topic branches that don't have PRs.

**Comment by kparzysz-quic** [comment]
Turns out there is already an [issue](https://github.com/llvm/llvm-project/issues/56637) opened for merging policy.

**Comment by rengolin** [comment]
> I propose an amendment to your idea, that we only delete (via scripts) topic branches that don't have PRs.

That's what I meant.

**Comment by rengolin** [comment]
> The problem with only allowing topic branches when they are used in chained PRs is that once all but 1 PR remains unmerged, the remaining PR is no longer "stacked". In that sense, a "standalone" PR is a special case of chained PRs.

This is a special case and it's perfectly fine to have that. 

I wager that even having it as a means to create a single shared PR between two or more developers would be fine, as long as the branch is deleted when the PR is either merged or abandoned, and that we constantly clean up left over PRs.

This can look very similar to feature branches, but we have strong policies for large PRs, so we can restrict the amount of work that is done in a branch similar to what developers would do alone in their forks (including rebase, cleanups, etc).

**Comment by kparzysz-quic** [comment]
How should be deal with abandoned PRs?  When can we declare a PR as abandoned?  Should we define a period of inactivity as evidence of abandonment?

**Comment by rengolin** [comment]
> 1. Allow developers to create topic branches in llvm-project. The downside of the first option is that this can lead to hundreds of leftover branches, which is messy. This should also include policy on deleting stale (or otherwise unneeded) branches.
> 2. Do not allow topic branches, and only allow cross-fork pull requests. The downside of the second option is that it would prevent chaining PRs. This may or may not be a big deal, depending on whether we decide to use chained PRs or not.

3. Only allow topic branches if they're tied to PRs, which must be set to delete the branch when merged/abandoned. (see #56637). This has none of the problems above. :)


> First of all, there is still the question of users merging branches into main directly, without GitHub. The move to PRs may seem to imply that direct pushes are no longer allowed, but I wanted to address that explicitly.

Ah, yes, that did occur to me reading this but I think it's orthogonal. Also, I don't think we should change too many policies in one go.

As it stands, people can push directly (code owners, fixups, etc). People can create branches directly, but we currently strongly advise against. The discussion here, as I see, is if we'll enforce that as standard and possibly automated.

If we have the script deleting branches that are not associated with PRs, then it should work for all cases.


> 1. Create a merge commit:  this option makes the commit history non-linear.  If we follow the pre-existing policy of keeping the history linear, we should disallow it.

Never.

> 2. Rebase and merge: rebase the topic branch associated with the PR on top of the base branch of the PR. Every commit in the PR branch will be present as a separate commit in the base branch.
> 3. Squash and merge: squash all commits in the PR branch into one commit and cherry-pick to the base branch.

Either of these two are fine with current LLVM policies. It's up to the community and the discussion in #56637 whether we want one of those or allow either.

**Comment by rengolin** [comment]
> When can we declare a PR as abandoned? Should we define a period of inactivity as evidence of abandonment?

That's a very good question.

Abandoning a PR doesn't delete the review, which can be picked up later and even restarted. However, it would (in my proposal) delete the branch ONLY IF it's a branch in the main repo. People can have their branches in their own forks.

This would encourage people to develop in their own forks, but would be a problem for abandoning stacked PRs.

I guess if the policy is to warn people the PR has been left unattended for too long before abandoning / deleting stuff, leaving time for them to backup on their forks or whatever, it would be fine. But the timing must be very lenient (months), as people do go on long holidays, sabbaticals, health issues that can take a long time.

Worst case (can't contact the author and the branch is in the main repo) we keep the PR open? Or give an even longer time? There's some bike-shedding to be done there, for sure. :)

**Comment by nhaehnle** [comment]
One really important aspect here is: even if branches in the main repository are allowed, who is allowed to create them?

If it is restricted to people who can currently push to main, then by definition it's an exclusionary policy to newcomers. One stated goal of this entire project is to make LLVM more approachable, so there's a direct conflict there.

On the other hand, opening the creation of branches to everybody seems like a scary idea if it is even possible in the first place, because of spam, vandalism, other forms of malice.

**Comment by rengolin** [comment]
> If it is restricted to people who can currently push to main, then by definition it's an exclusionary policy to newcomers. One stated goal of this entire project is to make LLVM more approachable, so there's a direct conflict there.

There is no conflict here. No sane Github repository allows just about anyone to create branches on its projects. Doing that would be maintenance madness. LLVM is really the outlier in allowing so many people commit to its repository directly.

We rely on people understanding their boundaries (which they do). In case someone doesn't (by either accident or malice), we can always restore the state from numerous mirrors and forks and deal with the incident on a case by case basis.

The use of Github may allow us to revisit the liberal commit policy, however. Since we request reviews and approving a PR means it can be automatically merged, we need less people with those permissions, not more.

**Comment by nhaehnle** [comment]
It seems we're in agreement? The conflict is that if there is a workflow (like stacked PR) that requires creating branches in the main repository, then this turns newcomers very explicitly into secondary citizens, which conflicts one of the stated goals of using GitHub PRs (attracting new contributors).

As a corollary, it looks to me like we have to find a solution for stacked reviews / multi-commit reviews that works **without** branches in the main repository.

**Comment by rengolin** [comment]
> As a corollary, it looks to me like we have to find a solution for stacked reviews / multi-commit reviews that works without branches in the main repository.

Or in a way that doesn't need users to **create** branches themselves, but are still able to use them for the stacked PR. Which admittedly, can easily become unwieldy.

One thing I don't like about stacked PRs as branches is that, reasonably often, we ask people to split a patch half-way, so for example, 7/10 become 7/11 + 8/11 and the branches afterwards are renamed accordingly.

So...
* If the branch naming in the PR is something like `my_pr_1`, `my_pr_2`..., we'll have a bad renaming problem;
* If we ask people to name each branch, we could have a name clash problem;
* If we want the branch to be PR#, we need to get the PR first, which we can't, because we don't have a branch;
* If a branch is left out of a PR by accident, or during a move, and we run the cronjob, it could wrongly "clean it up";
* Etc...

To me, stacked PRs using branches is really fragile. I personally think we should just use a single PR with multiple commits and allow people to rebase and merge at the end. I know Github isn't great to review multiple commits in a single PR, but it's even worse to do PRs on top of PRs using multiple branches.

To be honest, Phab isn't great for stacked PRs either, and I do review quite a lot of those (every new target). There's a lot of manual labour, which often goes wrong because the people submitting are new to LLVM (and Phab), and because even I don't remember all the rules every time.

**Comment by nikic** [comment]
I don't think it is viable to allow branches on the main repo. The monorepo is already stretching the bounds of what is reasonable (see e.g. the massive pushback the recent CIRCT integration proposal received, nearly entirely for general monorepo concerns), and allowing people to push their work-in-progress patches to the monorepo will make things even worse.

**Comment by rengolin** [comment]
Another problem of using the main repo's branches as stacked PR (or general development) is that we allow _everyone_ to create branches (write permission), thus we also allow _everyone_ to delete/rename branches on the main repo.

As examples, on other Github projects:
* I have accidentally deleted my own PR's branch, thinking it was a stale branch, which then closed the PR;
* I have accidentally clicked on the wrong "bin icon" and deleted the branch immediately above/below the one I actually wanted;

Both cases above can disrupt in-flight reviews, pushes and other active work on the branch.

Luckily, on all cases above, either me or someone else noticed the problem and we could correct in time (before Github cleaned up deleted branches), but there weren't many active developers. 

I don't think this scales well to the number of _writers_ we have in our repo.

**Comment by kparzysz-quic** [comment]
I think either policy (either allowing or disallowing user-created branches in the main repo) makes sense, but it seems like disallowing it comes with a lot less pitfalls.

**Comment by rengolin** [comment]
> but it seems like disallowing it comes with a lot less pitfalls.

Only if we can work stacked PRs without them.

**Comment by nikic** [comment]
> > but it seems like disallowing it comes with a lot less pitfalls.
> 
> Only if we can work stacked PRs without them.

I'm not sure that's really a requirement. Stacked reviews are a Phabricator paradigm that does not translate well to GitHub. I don't think we need an exact equivalent for it, just something that serves the same general purpose.

I believe in many cases, we'd want to replace stacked reviews with a single PR with multiple commits. It's still possible to review the commits individually. A reviewer can indicate that some subset of the commits looks good, and those can be landed separately.

If just having multiple commits does not work well for some case, the usual pattern I've seen is that people just submit a PR that contains another PR as a subset and just leave a comment to ignore everything but the last commit, or similar. As GitHub allows you to review individual commits in a PR, this is workable.

**Comment by nikic** [comment]
To give an example, a common use-case for stacked reviews by new contributors is to pre-commit tests in a separate review (as they can't pre-commit to the repo). Having a single PR with two commits, one adding the tests and one implementing the change, serves that use-case much, much better than stacked PRs (or even Phabricator's stacked reviews, which add a lot of additional friction for this simple use-case).

**Comment by rengolin** [comment]
> I believe in many cases, we'd want to replace stacked reviews with a single PR with multiple commits. It's still possible to review the commits individually. A reviewer can indicate that some subset of the commits looks good, and those can be landed separately.

Ah, I see! We're talking about the same thing, then.

When I say "stacked PR", I mean the thing we do in Phabricator for multiple commits. I would _also_ prefer if this is just done as a PR with multiple commits here, but some people have expressed concerns on how to conduct the review, because Github is quite bad at it when you start shuffling commits around, removing, force-pushing, etc.

But I think if we agree with, for example, no force pushing and a merge push at the end (with appropriate commit message editing), then it should be fine, IMO.

**Comment by nhaehnle** [comment]
> As GitHub allows you to review individual commits in a PR, this is workable.

What, specifically, do you mean by this? I'm aware of the ability to select individual commits in the "Files changed" tab of PRs. Is there anything else?

One of the biggest gaps is that AFAIK GitHub tracks Accept/Reject for the PR as a whole and not for individual commits.

It would be great to have some mechanism to do this and perhaps even automatically add "Reviewed-by" (and "Acked-by"?) footers. Maybe a bot?

**Comment by bryanpkc** [comment]
> One of the biggest gaps is that AFAIK GitHub tracks Accept/Reject for the PR as a whole and not for individual commits.

In GitHub workflow, if a reviewer accepts some but not all of the commits in a PR, he would have to "Request Changes" on the PR, specifying what is acceptable and what is not. The author can then force-push the branch to eliminate the commits that are not yet ready (and can submit new PRs with rewritten versions of the same). It's quite clumsy.

**Comment by bryanpkc** [comment]
As an author, I personally like to force-push the PR after reviews are done, so that I can squash the commits and reword as appropriate, but GitHub's (current) poor UI for displaying and comparing historical force-pushes makes life difficult for reviewers and future readers of the PR.

**Comment by rengolin** [comment]
> One of the biggest gaps is that AFAIK GitHub tracks Accept/Reject for the PR as a whole and not for individual commits.

We already require that all commits in a stacked review to be approved before merge, I don't see this as a strong requirement.

> It would be great to have some mechanism to do this and perhaps even automatically add "Reviewed-by" (and "Acked-by"?) footers. Maybe a bot?

Adding reviewed by would be interesting (and it's a different concern), but if we merge with the #PR number, then we don't need to track who approved, as the PR already shows it.

**Comment by rengolin** [comment]
> As an author, I personally like to force-push the PR after reviews are done, so that I can squash the commits and reword as appropriate, but GitHub's (current) poor UI for displaying and comparing historical force-pushes makes life difficult for reviewers and future readers of the PR.

I'm the same with force-pushes, but I got used to not doing that on my side and just squash merge and editing the commit message in the UI. Avoids an extra run of the pre-merge CI and uses the web interface that is similar to the console interface that git+vim provides, so...

**Comment by jyknight** [comment]
The commit message is something that should be reviewed too -- just as the
code is. A system where you can easily write it in advance and iterate on
it is a major plus, and generally leads to much better commit messages.

Going back to having the commit message be something which can only be
written immediately before pushing means that it likely won't be very well
written, and that it cannot be reviewed. That seems unfortunate.

**Comment by bryanpkc** [comment]
> The commit message is something that should be reviewed too

To be fair, on GitHub we can have a policy of not squash-merging, but requiring the author to squash and reword their own commits before a committer merges the PR. The committer thus has a chance to review the commit message, and request rewrites if necessary (which unfortunately will require another force-push). But the UI discourages the reviewers from doing so, by hiding the commits in another tab, and requiring additional mouse clicks to expand them.

**Comment by hubert-reinterpretcast** [comment]
> If just having multiple commits does not work well for some case, the usual pattern I've seen is that people just submit a PR that contains another PR as a subset and just leave a comment to ignore everything but the last commit, or similar. As GitHub allows you to review individual commits in a PR, this is workable.

Until you need to add changes to respond to comments on the PR (then you want people to review a specific commit range -- or you start amending/force pushing).

Or it's works until you need to merge changes that happened "upstream" from that PR (I'm not sure how that case works except to use an intermediate base branch -- or you start rebasing/force pushing).

Not friendly without force pushes, not friendly with force pushes.

**Comment by jh7370** [comment]
As a user of GitHub within our work environment, I've found that PRs with multiple commits that are intended for individual review has never worked well. The GitHub UI is completely unintuitive for this, I and others find, and this often leads to confusion. This might be due to unfamiliarity of the reviewer with GitHub admittedly, but if the intent is to get new contributors _reviewing_ as well as committing, we no doubt will have similar problems. Note that this is just from the "diff" panel of the PRs too - what happens in the conversation view is an even worse story, but probably out-of-scope for this discussion.

Aside from the uninituitive UI (which could always be improved), I don't see how stacked commits within a single PR can work with potential fix-ups to commits somewhere in the range, without force pushes and the consequent destruction of comments on subsequent commits in the series, or without those fix-ups appearing in the wrong place in the commit sequence. A final interactive rebase to squash in fix-ups is one solution to this, although again I found that in practice people not particularly familiar with git get thrown by this.

Another issue with stacked commits within a single PR is making sure reviewers are appropriately selected - you want to be able to precisely target the added reviewers and subscribers so that they don't have to wade through the noise of a large PR to find the bit that touches the one file they are a maintainer of.

All that being said, I do agree that having development branches in the main repo is not a great idea, without some method of auto-cleaning them up, and even then there are problems with it: every branch in existence adds to the amount of metadata in the git repo, increasing the repo size, clone/fetch times etc, and since I expect there to be several hundred branches on the go at any one time (quite possibly more), this will quickly add up to an (even more) unmanageable repository size. On the other hand, I also dislike the branch from fork approach because it feels a little heavy-weight to me (in theory - I've never actually had a need to try it out myself).

Basically, I don't see an approach that works well and involves using GitHub...

**Comment by joker-eph** [comment]
Stacked PRs involved frequently different components which targets different reviewers. I don't quite get how to keep separate thread of reviews with different reviewers in the same PR. 
> Then you want people to review a specific commit range -- or you start amending/force pushing).

It is worth than a range, it becomes a list of non-contiguous commit, because you push the "address comment" commits for all the previous ones on top of the branch.

> One thing I don't like about stacked PRs as branches is that, reasonably often, we ask people to split a patch half-way, so for example, 7/10 become 7/11 + 8/11 and the branches afterwards are renamed accordingly.

The branches don't strictly need to be named, one can also use a naming scheme that would look like: `branch/my-feature/improve-lib-support`, `branch/my-feature/add-base-class`, `branch/my-feature/define-interface`, ...
If the ordering is really important you can use sub numbering: `branch/my-feature/1`,  `branch/my-feature/2` ... and when you split the number 7 you create a new  `branch/my-feature/7b` that fits well in the middle.

That said I am not sure how to allow branches in the repo in a way that won't be messy and hard to maintain, maybe the scripting and auto-cleanup may be maintainable? Do we have experience with this from other large GitHub projects?

> It seems we're in agreement? The conflict is that if there is a workflow (like stacked PR) that requires creating branches in the main repository, then this turns newcomers very explicitly into secondary citizens, which conflicts one of the stated goals of using GitHub PRs (attracting new contributors).

It isn't clear to me that we **have** to get a first time contributor up and running with stacked pull requests immediately: it is really unreasonable to consider these a more "advanced" feature? People get LLVM commit access fairly quickly/easily right now.

>  even then there are problems with it: every branch in existence adds to the amount of metadata in the git repo, increasing the repo size, clone/fetch times etc

Can we use a branching namespace that isn't part of the clone? Could this be basically "free"?

**Comment by joker-eph** [comment]
Maybe the entire "should we allow branches in the repo" discussion is really something that should be settled in https://github.com/llvm/llvm-project/issues/56636 ? If the requirements come up with "branches are needed" then we should work on the policy and tooling to make it manageable and scalable without an important cost (clone size for example)

**Comment by rengolin** [comment]
Branches in the main repo are more than just a means to an end. They have serious security issues that we cannot overlook.

As I said elsewhere: requiring branches on the main repo for new reviews will require people to have write access _before_ having contributed to LLVM at all, _before_ we can even see what the pull request is about.

This is a really easy attack vector for people/groups trying to undermine the users of a "very popular compiler architecture" or just script kiddies trying to prove a point or show off to others. If they do it right (and slow), it would be really hard to separate who's a real new developer and who's just a dormant account waiting to create chaos.

This isn't a problem in closed projects, because the restriction that a person needs to be from a company, considerably reduces that risk. But with the drive to bring in more people into LLVM, we could very easily not realise we're letting in really good social engineers.

**Comment by rengolin** [comment]
Creating a pull-request from a fork doesn't need the user to have write permission on the destination repository, while pushing changes to a branch does. It's all about permissions.

From a fork, it's GH who adds the (implicit) branch, not the final user.

**Comment by joker-eph** [comment]
What is the practical difference? You claimed a "security" concern and I was trying to understand why it matters whether the branch ends up pushed in the repo from the command line or from a click in a fork in the UI?

**Comment by rengolin** [comment]
It's not about UI, it's about permission. I'm not sure how to be more clear.

The security concern is that, if we give write permissions to new developers _before_ they show they're serious (against our current policy), they now have the write to do anything they want, including direct pushes, approving/rejecting other people's PRs, merging incomplete reviews, etc.

**Comment by joker-eph** [comment]
OK it wasn't clear to me that you were concerned about something else than "pushing a branch". I asked in the other thread if "groups ACL" aren't allowing to protect `main` (folks can't touch it) while allowing them to push a user branch?  (That is "write permission" isn't necessarily an "all or nothing").

Also: we may be able to **allow** user branch without **requiring** it, allowing contributions from forks in general.

**Comment by rengolin** [comment]
Write permission gives people the right to approve/reject/merge PRs and edit other things in the repo.

There are many ways in GH to protect different things, but the more you deviate from the beaten path, the harder it gets to get it right and do it consistently. I have to manage a few GH projects and each as a slightly different way of handling permissions.

I personally think we should be reducing the number of people with write permission, not increase. For example, trimming people that haven't contributed recently (code, reviews, etc) for a set period of time, but certainly avoiding giving new people write access after a few merged PRs.

People who need write access are those that actively work on LLVM. Reviewers, main coders, active code owners, testers, release people, infrastructure, etc. Many of the people with write access today haven't touched LLVM in a long time or haven't gone further than submitting a few patches when they needed and moved on.

**Comment by nhaehnle** [comment]
I wonder if stacked PRs could be made to semi-work without branches in the main repo if we had a bot that automaticallý edited the initial comment of all the PRs to add a table of all PRs in the stack?

(I'm on the fence as to whether stacked PRs or PRs with multiple commits are better. I feel we sort of need to explore both in parallel to see how well they can be made to work.)

**Comment by joker-eph** [comment]
> if we had a bot that automaticallý edited the initial comment of all the PRs to add a table of all PRs in the stack?

Maybe I miss something I think you're trying to address the part about navigating the stack of PRs. 
However the main pain point for reviewing a stack of PRs is rather that you want each PR to show only the changes from this PR and not the one from the previous ones in the stack: this is what chaining them in the UI brings, it hides the commits from the "parent" PRs by having them target each others in cascade.

**Comment by joker-eph** [comment]
> Can PRs in stacked PRs be "retargeted" to a different branch, to re-order a stack?

Yes, I know I can do this manually, and I suspect https://github.com/ejoffe/spr does it automatically, you keep your linear sequence of commits locally, reorder them so that the one ready to land is the very first one, and from there you just sync the PRs to this state and you can merge.

You can also cherry-pick the commit on top of main, add "Closes #1234" and push, it'll close the PR 1234. If it was the 3rd PR in the chain you can edit in the UI the 4th PR to target the branch for the second PR, skipping the merged one in the cascade/chain.

**Comment by nhaehnle** [comment]
> > if we had a bot that automaticallý edited the initial comment of all the PRs to add a table of all PRs in the stack?
> 
> Maybe I miss something I think you're trying to address the part about navigating the stack of PRs.

Exactly. I was thinking about what it would take to make GitHub's stacked PR feature usable with branches in forks (in which case, your main pain point is the one (and only!?) thing that GitHub solves for us...).

**Comment by hubert-reinterpretcast** [comment]
> Exactly. I was thinking about what it would take to make GitHub's stacked PR feature usable with branches in forks (in which case, your main pain point is the one (and only!?) thing that GitHub solves for us...).

I think that messes with user notifications if some of the PRs are in the fork (not the main project).

**Comment by nhaehnle** [comment]
The question of whether branches should be allowed in he upstream repository came up again in a Discourse discussion (["[RFC] Revisiting linear history vs merge commits"](https://discourse.llvm.org/t/rfc-revisiting-linear-history-vs-merge-commits/64873/22)).

**Comment by tstellar** [comment]
It seems to me that the consensus here is that for either general development or non-stacked pull requests, user branches should not be allowed in the main repo, but whether or not to allow user branches for stacked pull requests is still undecided.  

Or to state it another way, we've narrowed this down to two options:

1. Don't allow any user branches in the main repo.
2. Only allow user branches for stacked pull requests.

Does this seem like a fair summary?

**Comment by jh7370** [comment]
I'm fine with either approach (as long as we have a stacked PR solution anyway), but we should make sure there is adequate documentation for how to create a fork and reference it in an LLVM PR.

**Comment by rengolin** [comment]
Can you describe how we'll "only allow user branches for pull requests"? Even before considering the security implications, there's no setting that separates it. I don't know how you could, and if that's true, it's not really an option, per se.

**Comment by tstellar** [comment]
@rengolin We can at least have a policy to not create branches for pull requests even if we can't enforce it technically.  Then at least people won't be creating them on purpose, so we won't get as many. I do think there still may be some solutions for technical enforcement, but we need to this investigate this more.

**Comment by tstellar** [comment]
> I've submitted a documentation change based on the summary of this discussion: https://reviews.llvm.org/D147276

This patch has been merged, so I'm going to close this issue.
